File: CN/CRC/crc.java

#include <iostream>

using namespace std;

class crc{
    public:
    
    int data_bits, divisor_bits, dataword[20], divisor[20], temp[20];
    
    void getData(){
        cout << "***Connected with receiver enter data to send***\n";
        cout << "Enter Dataword size : ";
        cin >> data_bits;
        
        cout << "Enter Divisor size : ";
        cin >> divisor_bits;
        
        cout << "Enter dataword : ";
        for(int i=0; i<data_bits; i++){
            cin >> dataword[i];
        }
        
        cout << "Enter divisor : ";
        for(int i=0; i<divisor_bits; i++){
            cin >> divisor[i];
        }
        
        divisor_bits--;
        
        
         for(int i=0;i<divisor_bits;i++)
            dataword[data_bits+i] = 0;
            
            
        for(int i=0;i<data_bits+divisor_bits;i++)
            temp[i] = dataword[i];
            
        exor_divide(temp, divisor, data_bits, divisor_bits);
    }
    
    
    
    
    void display(){
        cout << "dataword : ";
        for(int i=0; i<data_bits; i++){
            cout << dataword[i];
        }
        
        cout << "divisor : ";
        for(int i=0; i<divisor_bits; i++){
            cout << divisor[i];
        }
    }
    
    void exor_divide(int temp[],int divisor[],int data_bits,int divisor_bits)
        {
         for(int i=0;i<data_bits;i++)
         {
             if (divisor[0]==temp[i])
             {
                 for(int j=0,k=i;j<divisor_bits+1;j++,k++)
                     if(!(temp[k]^divisor[j]))
                         temp[k]=0;
                     else
                         temp[k]=1;
             } 
         }
        }
        
        
    void send_message(){
         cout<<"Genaerated CRC : ";
             for(int i=0;i<divisor_bits;i++)
             {
                 cout<<temp[data_bits+i]<<" ";
                 dataword[data_bits+i] = temp[data_bits+i];
             }
             
             cout<<endl<<"Sending  Message  as : ";
             for(int i=0;i<data_bits+divisor_bits;i++){
                 
                 cout<<dataword[i]<<" ";
             }
    }
    
    void receiver(){
        cout << endl << endl;
        cout << "***Receiving data***"<<endl;
        
        cout<<"Enter the message received to you to check error : ";
         
         for(int i=0;i<data_bits+divisor_bits;i++){
             cin>>dataword[i];
         }
         
             
         for(int i=0;i<data_bits+divisor_bits;i++)
             temp[i] = dataword[i];
             
         exor_divide(temp,divisor,data_bits,divisor_bits);
         
         for(int i=0;i<divisor_bits;i++)
         {
             if(temp[data_bits+i])
             {
                 cout<<"**Message is discarded due to error**";
             }
        }
        
         cout<<"Decoded successfully no error found** : ";
         cout << "Message Send by receiver : " << endl;
         for(int i=0; i<data_bits; i++){
             
             cout<<dataword[i]<<" ";
             
         }
    }
    
};


int main()
{
    crc obj;
    
    obj.getData();
    
    obj.send_message();
    
    obj.receiver();
    
    return 0;
}

File: CN/CRC/crc.cpp
#include <iostream>
using namespace std;

class crc{
public:

int dataword_size, divisor_size, dataword[20], divisor[20], temp[20], codeword[20];
int flag=0;

void getData(){
cout << "Enter Dataword size : ";
cin >> dataword_size;

cout << "Enter Divisor size : ";
cin >> divisor_size;

cout << "Enter dataword : ";
for (int i = 0; i < dataword_size; i++)
{
    cin >> dataword[i];
}

cout << "Enter Divisor : ";
for (int i = 0; i < divisor_size; i++)
{
    cin >> divisor[i];
}

cout << "Original dataword : ";
for (int i = 0; i < dataword_size ; i++)
{
   cout << dataword[i];
}


// append divisor-1 0's to dataword

divisor_size--;

for (int i = 0; i < divisor_size; i++)
{
    dataword[dataword_size+i] = 0; 
}

cout << "\nDataword  after padding : ";
for (int i = 0; i < dataword_size+divisor_size ; i++)
{
   cout << dataword[i];
}

// set temp to dataword
for (int i = 0; i < dataword_size+divisor_size ; i++)
{
   temp[i] = dataword [i];
}

cout << "\nTemp value : ";
for (int i = 0; i < dataword_size+divisor_size ; i++)
{
   cout << temp[i];
}

// call crc method

ex_or(temp,divisor,dataword_size,divisor_size);

}

void ex_or(int temp[],int divisor[],int dataword_size,int divisor_size){

for (int i = 0; i < dataword_size; i++)
{
   if (divisor[0] == temp[i])
   {
    for (int j = 0, k=i; j < divisor_size+1; j++,k++)
    {
        if (!(temp[k]^divisor[j]))
        {
            temp[k] = 0;
        }
        else{
            temp[k] = 1;
        }
        
    }
    
   }
}
}

void sender_side(){
    cout << "\nGenerated CRC : ";
    for (int i = 0; i < divisor_size; i++)
    {
        cout << temp[dataword_size+i];
        dataword[dataword_size+i] = temp[dataword_size+i];
    }

    cout << "\nGenerated codeword to send it to receiver : ";
    for (int i = 0; i < dataword_size+divisor_size; i++)
    {
       cout << dataword[i];
    }
    

}

void receiver_side(){
    cout << endl << endl;
    cout << "Enter the codeword you received : ";
    for (int i = 0; i < dataword_size+divisor_size; i++)
    {
        cin >> codeword[i];
    }

    ex_or(codeword,divisor,dataword_size,divisor_size);

    for (int i = 0; i < divisor_size; i++)
    {
        if (codeword[dataword_size+i])
        {           
            flag = 1;
            break;
        }
    }
       // cout << "Decoded successfully!";
       if (flag==1)
       {
        cout << "Error in transmission !";
       }
       else{
        cout << "Correct message is received!";
       }
         
}

};

int main(){

    crc obj;

    obj.getData();

    obj.sender_side();

    obj.receiver_side();

    return 0;
}
